chirp/drivers/radioddity_gm30.py

```diff
@@ -0,0 +1,967 @@
+    # obtained by tlmap query. each loc has 4 (pre)fix
+    # block numbers. Each block is 64 bytes
+    data = b''
+    for i in range(16):
+        for pre in range(0x0, 0xc1, 0x40):
+            addr = struct.pack('>BBBB', pre, loc + i, 0x00, 0x40)
+            data += do_read_cmd(serial, addr, 0x40)
+            status.cur += 0x40
+            radio.status_fn(status)
+    return data
+
+
+def do_upload_block(serial, loc, offset, radio, status):
+    # write each config type to loc retrieved from tlmap
+    # 4 defined (pre)fixes/blocks per (loc)ation
+    # pre = 0x00 , 0x40 , 0x80, 0xc0
+    # full command example  [57][80][4d]0040
+    # 57 = write , 80 prefix/block , 4d location to write
+    _mem = radio._memobj.get_raw()
+    for i in range(16):
+        for pre in range(0x0, 0xc1, 0x40):
+            x = offset + (i * 0x100) + pre
+            block = _mem[x:x + 0x40]
+            addr = struct.pack('>BBBBB', 0x57, pre, loc + i, 0x00, 0x40)
+            data = addr + block
+            ack = raw_send(serial, data, 1)
+            if ack != b'\x06':
+                err = f"Bad ack on write expect 0x06 got {ack}"
+                LOG.debug(err)
+                raise errors.RadioError(err)
+            status.cur += len(block)
+            radio.status_fn(status)
+
+
+def do_download(radio):
+    data = b''
+    try:
+        status = chirp_common.Status()
+        serial = radio.pipe
+        serial.flush()
+        ident = enter_prog(serial)
+        check_ident(ident)
+        do_sysinfo(serial)
+        do_readconfig(serial)
+        do_prog2(serial)
+        tl_map = do_read_tlmap(serial)
+        status.max = len(tl_map) * 0x1000
+        status.msg = "Downloading..."
+        for t, loc in tl_map.items():
+            if loc == 0:
+                raise errors.RadioError(f"TL Map failed {t} {loc}")
+            data += do_read_ranges(serial, loc, radio, status)
+    except Exception as e:
+        raise errors.RadioError(f"Error during download {e}")
+    finally:
+        exit_prog(serial)
+    return memmap.MemoryMapBytes(data)
+
+
+def do_upload(radio):
+    try:
+        status = chirp_common.Status()
+        status.max = len(WRITE_MAP) * 0x1000
+        serial = radio.pipe
+        serial.flush()
+        ident = enter_prog(serial)
+        check_ident(ident)
+        do_sysinfo(serial)
+        do_readconfig(serial)
+        do_prog2(serial)
+        tl_map = do_read_tlmap(serial)
+        for wt, label, offset in WRITE_MAP:
+            status.msg = f"Uploading: {label}..."
+            loc = tl_map[wt]
+            do_upload_block(serial, loc, offset, radio, status)
+    except errors.RadioError:
+        raise
+    except Exception as e:
+        raise errors.RadioError(f"Error during upload {e}")
+    finally:
+        exit_prog(serial)
+
+
+@directory.register
+class RadiodditysetGM30(chirp_common.CloneModeRadio):
+    """Radioddity GM30"""
+    VENDOR = "Radioddity"
+    MODEL = "GM30"
```

Comment:
Looks like this should be "GM-30" like all their other models and their marketing material: https://www.radioddity.com/products/radioddity-gm-30